### PR TITLE
feat: Support homedir expansion in lazy/scan read functions

### PR DESCRIPTION
**Pull request by alexander-beedie**
(@ritchie46: as mentioned yesterday)

We already expand the homedir `~/` for eager read function paths, but seem to have missed it for the lazy/scan equivalents; this just adds the missing expansion there too.

### CODE REVIEW

File: crates/polars-lazy/src/scan/ipc.rs [hunk 1]
```diff
@@ -96,11 +97,17 @@ impl LazyFileListReader for LazyIpcReader {
...
+        self.path = resolve_homedir(&path);
         self
     }
 
     fn with_paths(mut self, paths: Arc<[PathBuf]>) -> Self {
+        let paths = paths
+            .iter()
+            .map(|p| resolve_homedir(p))
+            .collect::<Vec<_>>()
+            .into();
```
Comment (ritchie46) [inline crates/polars-lazy/src/scan/ipc.rs:109]: Not entirely sure, but I believe we can collect into a `Arc<[]>` directly.

Comment (alexander-beedie) [reply crates/polars-lazy/src/scan/ipc.rs:109]: Ahhh; think an `into_iter()` with a `&p` should optimise appropriately.
Updated 🤔

Comment (alexander-beedie) [reply crates/polars-lazy/src/scan/ipc.rs:109]: Hmm, can't see a way to avoid the intermediate `Vec`; the `Arc` needs a constructed object to reference 🤔 If you can spot something cunning, feel free to point me at it or commit over the top :))

**Comment by etiennebacher** [comment]
@alexander-beedie Just wondering if `resolve_homedir()` should also be used for `write_` functions? For context, we want to fix a bug on this in `r-polars` (https://github.com/pola-rs/r-polars/issues/1085) and we're not sure if the path expansion should be done in Rust or in R. I see that `write_csv()` calls `file = normalize_filepath(file)` in Python but with this PR this is handled in Rust (for read/scan functions).

Bottom line: if the user calls `.write_csv("~/foo.csv")`, should this path be passed directly to Rust or should it be modified before?

File: crates/polars-lazy/src/scan/ipc.rs (hunk 1 above)
Comment (ritchie46) [reply crates/polars-lazy/src/scan/ipc.rs:109]: Now I am sure. :D

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=82ca9fa7f2c267ad4811d7714747b7ae

Comment (alexander-beedie) [reply crates/polars-lazy/src/scan/ipc.rs:109]: > Now I am sure. :D
> 
> https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=82ca9fa7f2c267ad4811d7714747b7ae

Lol... pretty sure that the `collect` there still creates an intermediate `Vec`, it just gets deallocated after conversion to `Arc[i32]`  :)

Comment (ritchie46) [reply crates/polars-lazy/src/scan/ipc.rs:109]: Yes, but Rust might optimize that later.

**Comment by alexander-beedie** [comment]
(Need to check if this is still an issue; if it is will take a fresh look at it - just cleaning up some old PRs 😅)
